401 Unauthorized in Clusterized Environment (Replicas > 1)

### Current Behavior

It is not possible to apply our realm file as a 401 is returned by Keycloak when using `keycloak-config-cli` with more than one replicas in Keycloak. The issue happens when trying to retrieve `/serverinfo` (after successful login).

### Expected Behavior

Synchronisation should happen without error.

### Steps To Reproduce

```markdown
1. Run Keycloak in clusterised environment
2. Sync
```


### Environment

- Keycloak Version: 12.0.4
- keycloak-config-cli Version: 4.2.0-12.0.4


### Anything else?

Our Keycloak instances are running with the following environment variables set:

```
JGROUPS_DISCOVERY_PROTOCOL=JDBC_PING
JGROUPS_DISCOVERY_PROPERTIES=datasource_jndi_name=java:jboss/datasources/KeycloakDS,info_writer_sleep_time=500,initialize_sql="CREATE TABLE IF NOT EXISTS JGROUPSPING ( own_addr varchar(200) NOT NULL, cluster_name varchar(200) NOT NULL, created timestamp default current_timestamp, ping_data BYTEA, constraint PK_JGROUPSPING PRIMARY KEY (own_addr, cluster_name))"
```

Downgrading from 2 replicas to 1 makes it work again.
No logs is generated when trying to sync.

Thank you for your PR. I had the opportunity to test it and still am getting the same error, however this time with an additional warning:
```
2021-10-02 06:38:53.663  WARN 1 --- [           main] o.a.h.c.protocol.ResponseProcessCookies  : Invalid cookie header: "Set-Cookie: AWSALB=5Pb9h3xYOuuVx2X3BJIXuXNJ79Km+2atYUaJ2oPOt4YRUrzu3jqxjbuEaemo2ibDnow9XeTee/JpQ+ipWXIWGX9DIS3GxoT48hZjYUIX3LCj5Ah4PAQufhyRgltX; Expires=Sat, 09 Oct 2021 06:38:53 GMT; Path=/". Invalid 'expires' attribute: Sat, 09 Oct 2021 06:38:53 GMT
```
That for every requests that are made. I assume the cookie is still being discarded because of this warning. After digging a little bit it appears that the issue is linked to `expires` being a recent HTTP property that is not fully supported by default. I tried using a different HTTP Client but having issues compiling because of failing unit tests, I fear I reached the limits of my knowledge in Java.

src/main/java/de/adorsys/keycloak/config/util/ResteasyUtil.java:
```Java
RequestConfig requestConfig = RequestConfig.custom()
                .setCookieSpec(CookieSpecs.STANDARD)
                .build();
HttpClient apacheClient = HttpClientBuilder
                .create()
                .setDefaultRequestConfig(requestConfig)
                .build();
ApacheHttpClient43Engine engine = new ApacheHttpClient43Engine(apacheClient);
ResteasyClientBuilder clientBuilder = new ResteasyClientBuilderImpl();
clientBuilder
                .connectionPoolSize(10)
                .connectTimeout(10, TimeUnit.SECONDS)
                .readTimeout(10, TimeUnit.SECONDS)
                .httpEngine(engine)
                .build();
```
Unit tests fails with:
```
ERROR] Failures: 
[ERROR]   KeycloakProviderIT$HttpProxySpringProperties.run:100 
Expected: a string matching the pattern '.+ Connect to localhost:2 .+ failed: .+'
     but: was "RESTEASY004655: Unable to invoke request: java.net.UnknownHostException: keycloak: nodename nor servname provided, or not known"
[ERROR]   KeycloakProviderIT$HttpProxySystemProperties.testHttpProxy:86 
Expected: a string matching the pattern '.+ Connect to localhost:2 .+ failed: .+'
     but: was "RESTEASY004655: Unable to invoke request: java.net.UnknownHostException: keycloak"
[INFO] 
[ERROR] Tests run: 336, Failures: 2, Errors: 0, Skipped: 7
```

Thanks

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased]
 
+### Added
+
+- Cookie Management for http client to support clustered environments with cookie based sticky sessions
+
 ### Changes
 
 - Use java 17 as default and use docker image `openjdk:17-slim`

--- a/pom.xml
+++ b/pom.xml
@@ -251,6 +251,12 @@
             <version>${mockserver.version}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.mock-server</groupId>
+            <artifactId>mockserver-spring-test-listener</artifactId>
+            <version>${mockserver.version}</version>
+            <scope>test</scope>
+        </dependency>
 
         <dependency>
             <groupId>com.unboundid</groupId>

--- a/src/main/java/de/adorsys/keycloak/config/util/ResteasyUtil.java
+++ b/src/main/java/de/adorsys/keycloak/config/util/ResteasyUtil.java
@@ -20,6 +20,7 @@
 
 package de.adorsys.keycloak.config.util;
 
+import de.adorsys.keycloak.config.util.resteasy.CookieClientFilter;
 import org.jboss.resteasy.client.jaxrs.ResteasyClient;
 import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder;
 import org.jboss.resteasy.client.jaxrs.internal.ResteasyClientBuilderImpl;
@@ -57,6 +58,8 @@ public class ResteasyUtil {
                     Integer.parseInt(System.getProperty("http.proxyPort", "0"))
             );
         }
+
+        clientBuilder.register(new CookieClientFilter());
 
         return clientBuilder.build();
     }

--- a/src/main/java/de/adorsys/keycloak/config/util/resteasy/CookieClientFilter.java
+++ b/src/main/java/de/adorsys/keycloak/config/util/resteasy/CookieClientFilter.java
@@ -1,0 +1,52 @@
+/*-
+ * ---license-start
+ * keycloak-config-cli
+ * ---
+ * Copyright (C) 2017 - 2021 adorsys GmbH & Co. KG @ https://adorsys.com
+ * ---
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ---license-end
+ */
+
+package de.adorsys.keycloak.config.util.resteasy;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import javax.ws.rs.client.ClientRequestContext;
+import javax.ws.rs.client.ClientRequestFilter;
+import javax.ws.rs.client.ClientResponseContext;
+import javax.ws.rs.client.ClientResponseFilter;
+
+// By default, the RESTeasy cookie handling is very limited. There is no access to the underlying httpEngine.
+// The underlying httpEngine uses an old CookieSpec by default which is going to refuse valid cookie today.
+// Instead, build the httpEngine from scratch, we are using a RESTeasy filter to grab a re-attach cookie.
+// Currently, this filter does not valide cookie or is able to remove cookies.
+// A cookie managed is required to handle sticky sessions at cookie base
+public class CookieClientFilter implements ClientRequestFilter, ClientResponseFilter {
+    private final Map<String, String> cookies = new HashMap<>();
+
+    @Override
+    public void filter(ClientRequestContext clientRequestContext) throws IOException {
+        clientRequestContext.getHeaders().put("Cookie", new ArrayList<>(cookies.values()));
+    }
+
+    @Override
+    public void filter(ClientRequestContext requestContext, ClientResponseContext responseContext) throws IOException {
+        responseContext.getCookies().forEach((name, cookie) -> cookies.put(name, String.format("%s=%s",
+                cookie.toCookie().getName(),
+                cookie.toCookie().getValue()
+        )));
+    }
+}

--- a/src/test/java/de/adorsys/keycloak/config/mock/CookieMockIT.java
+++ b/src/test/java/de/adorsys/keycloak/config/mock/CookieMockIT.java
@@ -1,0 +1,106 @@
+/*-
+ * ---license-start
+ * keycloak-config-cli
+ * ---
+ * Copyright (C) 2017 - 2021 adorsys GmbH & Co. KG @ https://adorsys.com
+ * ---
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ---license-end
+ */
+
+package de.adorsys.keycloak.config.mock;
+
+import de.adorsys.keycloak.config.configuration.TestConfiguration;
+import de.adorsys.keycloak.config.extensions.GithubActionsExtension;
+import de.adorsys.keycloak.config.model.RealmImport;
+import de.adorsys.keycloak.config.provider.KeycloakImportProvider;
+import de.adorsys.keycloak.config.service.RealmImportService;
+import de.adorsys.keycloak.config.test.util.KeycloakMock;
+import org.junit.jupiter.api.MethodOrderer;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestMethodOrder;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockserver.client.MockServerClient;
+import org.mockserver.model.Cookie;
+import org.mockserver.springtest.MockServerTest;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.ConfigDataApplicationContextInitializer;
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.TestPropertySource;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import java.io.File;
+import java.io.IOException;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.hasSize;
+import static org.mockserver.model.HttpRequest.request;
+
+@ActiveProfiles("IT")
+@MockServerTest("keycloak.url=http://localhost:${mockServerPort}")
+@ExtendWith(SpringExtension.class)
+@ExtendWith(GithubActionsExtension.class)
+@ContextConfiguration(
+        classes = {TestConfiguration.class},
+        initializers = {ConfigDataApplicationContextInitializer.class}
+)
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+@ExtendWith(GithubActionsExtension.class)
+@TestPropertySource(properties = {
+        "import.state=false",
+        "import.force=true"
+})
+class CookieMockIT {
+    private MockServerClient client;
+
+    @Autowired
+    public KeycloakImportProvider keycloakImportProvider;
+    @Autowired
+    public RealmImportService realmImportService;
+
+    @Test
+    void run() throws Exception {
+        client.when(request().withPath("/realms/master/protocol/openid-connect/token")).respond(KeycloakMock::grantToken);
+        client.when(request().withPath("/admin/realms/simple")).respond(KeycloakMock::realmSimple);
+        client.when(request().withPath("/admin/realms/simple/default-default-client-scopes")).respond(KeycloakMock::emptyList);
+        client.when(request().withPath("/admin/realms/simple/default-optional-client-scopes")).respond(KeycloakMock::emptyList);
+        client.when(request().withPath("/admin/realms/simple/identity-provider/instances")).respond(KeycloakMock::emptyList);
+        client.when(request().withPath("/realms/master/protocol/openid-connect/logout")).respond(KeycloakMock::logout);
+
+        client.when(request().withPath("/admin/serverinfo")).respond((request) -> {
+            assertThat(request.getCookieList(), hasSize(2));
+            assertThat(request.getCookieList(), containsInAnyOrder(
+                    new Cookie("key_expires", "value"),
+                    new Cookie("key", "value")
+            ));
+
+            return KeycloakMock.serverInfo(request);
+        });
+
+        RealmImport realmImport = getRealmImport("import-files/simple-realm/00_create_simple-realm.json");
+        realmImportService.doImport(realmImport);
+    }
+
+    @SuppressWarnings("SameParameterValue")
+    private RealmImport getRealmImport(String file) throws IOException {
+        File realmImportFile = new ClassPathResource(file).getFile();
+
+        return keycloakImportProvider
+                .readRealmImportFromFile(realmImportFile)
+                .getRealmImports()
+                .get(realmImportFile.getAbsolutePath());
+    }
+}

--- a/src/test/java/de/adorsys/keycloak/config/test/util/KeycloakMock.java
+++ b/src/test/java/de/adorsys/keycloak/config/test/util/KeycloakMock.java
@@ -1,0 +1,91 @@
+/*-
+ * ---license-start
+ * keycloak-config-cli
+ * ---
+ * Copyright (C) 2017 - 2021 adorsys GmbH & Co. KG @ https://adorsys.com
+ * ---
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ---license-end
+ */
+
+package de.adorsys.keycloak.config.test.util;
+
+import org.mockserver.model.*;
+import org.testcontainers.shaded.com.fasterxml.jackson.core.JsonProcessingException;
+import org.testcontainers.shaded.com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.mockserver.model.HttpResponse.response;
+
+public class KeycloakMock {
+    static Headers cookieHeader = new Headers(
+            new Header("Set-Cookie", "key_expires=value; Expires=Sat, 09 Oct 2021 06:38:53 GMT; Path=/"),
+            new Header("Set-Cookie", "key=value")
+    );
+
+    public static HttpResponse grantToken(HttpRequest request) throws JsonProcessingException {
+        Map<String, Object> map = new HashMap<>();
+        map.put("access_token", "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c");
+        map.put("expires_in", 60);
+        map.put("refresh_expires_in", 1800);
+        map.put("refresh_token", "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c");
+        map.put("token_type", "Bearer");
+        map.put("not-before-policy", 0);
+        map.put("session_state", "a04afd4e-d871-4357-b635-1754138ba341");
+        map.put("scope", "profile email");
+
+        String json = new ObjectMapper().writeValueAsString(map);
+
+        return response()
+                .withHeaders(cookieHeader)
+                .withBody(json, MediaType.APPLICATION_JSON);
+    }
+
+    public static HttpResponse serverInfo(HttpRequest request) {
+        return response()
+                .withHeaders(cookieHeader)
+                .withBody("{\"systemInfo\":{\"version\":\"15.0.2\"}}", MediaType.APPLICATION_JSON);
+    }
+
+    public static HttpResponse realmSimple(HttpRequest request) throws JsonProcessingException {
+        if (request.getMethod("GET").equals("PUT")) {
+            return response().withStatusCode(204);
+        }
+
+        Map<String, Object> map = new HashMap<>();
+        map.put("id", "ef6addd5-7f2d-4781-ad89-99e0ed2e5eb7");
+        map.put("realm", "simple");
+        map.put("eventsEnabled", false);
+        map.put("attributes", new HashMap<>());
+
+        String json = new ObjectMapper().writeValueAsString(map);
+
+        return response()
+                .withHeaders(cookieHeader)
+                .withBody(json, MediaType.APPLICATION_JSON);
+    }
+
+    public static HttpResponse emptyList(HttpRequest request) {
+        return response()
+                .withHeaders(cookieHeader)
+                .withBody("[]", MediaType.APPLICATION_JSON);
+    }
+
+    public static HttpResponse logout(HttpRequest request) {
+        return response()
+                .withHeaders(cookieHeader)
+                .withStatusCode(204);
+    }
+}

--- a/src/test/resources/logback-test.xml
+++ b/src/test/resources/logback-test.xml
@@ -7,4 +7,5 @@
     <logger name="org.testcontainers" level="INFO"/>
     <logger name="com.github.dockerjava" level="WARN"/>
     <logger name="org.mockserver.log.MockServerEventLog" level="WARN"/>
+    <!--<logger name="org.apache.http.wire" level="DEBUG"/>-->
 </configuration>

